File: python/binary_tree_level_order_traversal_ii.py
# Definition for a binary tree node.
from typing import List
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def levelOrderBottom(self, root: TreeNode) -> List[List[int]]:
        if root is None:
            return []

        queue = deque()
        queue.append(root)
        result = []
        current_level = []

        while len(queue):
            current_level_length = len(queue)
            for _ in range(current_level_length):
                current = queue.popleft()
                current_level.append(current.val)
                if current.left:
                    queue.append(current.left)
                if current.right:
                    queue.append(current.right)
            result.append(current_level)
            current_level = []

        return result[::-1]

File: src/BinaryTreeLevelOrderTraversalII.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class BinaryTreeLevelOrderTraversalII {

    private static class Pair {
        TreeNode tree;
        int level;

        Pair(TreeNode tree, int level) {
            this.tree = tree;
            this.level = level;
        }
    }

    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> levels = new ArrayList<>();
        if (root == null) {
            return levels;
        }

        LinkedList<Pair> queue = new LinkedList<>();
        queue.add(new Pair(root, 0));
        List<Integer> level = new ArrayList<>();
        int currentLevel = 0;

        while (!queue.isEmpty()) {
            Pair pair = queue.pop();
            if (pair.level > currentLevel) {
                levels.add(level);
                level = new ArrayList<>();
                currentLevel = pair.level;
            }
            level.add(pair.tree.val);
            addChildrenToQueue(pair, queue);
        }
        levels.add(level);
        Collections.reverse(levels);
        return levels;
    }

    private static void addChildrenToQueue(Pair pair, Queue<Pair> queue) {
        if (pair.tree.left != null) {
            queue.add(new Pair(pair.tree.left, pair.level + 1));
        }
        if (pair.tree.right != null) {
            queue.add(new Pair(pair.tree.right, pair.level + 1));
        }
    }
}
